# CA1 - Task 2

We have been provided with the dataset, and the following model will be built using CRISP-DM framework. It stands for Cross Industry Standard Process for Data Mining, and it has 6 iterative phases.


1.   Business Understanding
2.   Data Understanding
3.   Data Preparation
4.   Modelling
5.   Evaluation
6.   Deployment

## Business Understanding


Bank wants to make use of a classification model which can predict whether the client has subscribed to a term deposit. The purpose of this classification model is to identify in the future whether a client will subscribe to a term deposit or not.

As per the business requirement, the evaluation metrics will be more focused on Recall and Precision, rather than Accuracy. Recall means the model will find those customers who are more likely to subscribe to a term deposit, and Precision means out of those customers, who will actually subscribe.

Both of these metrics will aid in optimising the bank's resources in marketing for term deposits.

## Data Understanding

In this section, we will collect and explore the dataset, to check its quality, gain initial insights and identify if there are any issues with the data.

In [55]:
from pandas import read_csv

In [56]:
## load the data
data = read_csv("/content/drive/MyDrive/ML_Dataset/bank.csv")
data

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,30,unemployed,married,primary,no,1787,no,no,cellular,19,oct,79,1,-1,0,unknown,no
1,33,services,married,secondary,no,4789,yes,yes,cellular,11,may,220,1,339,4,failure,no
2,35,management,single,tertiary,no,1350,yes,no,cellular,16,apr,185,1,330,1,failure,no
3,30,management,married,tertiary,no,1476,yes,yes,unknown,3,jun,199,4,-1,0,unknown,no
4,59,blue-collar,married,secondary,no,0,yes,no,unknown,5,may,226,1,-1,0,unknown,no
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4516,33,services,married,secondary,no,-333,yes,no,cellular,30,jul,329,5,-1,0,unknown,no
4517,57,self-employed,married,tertiary,yes,-3313,yes,yes,unknown,9,may,153,1,-1,0,unknown,no
4518,57,technician,married,secondary,no,295,no,no,cellular,19,aug,151,11,-1,0,unknown,no
4519,28,blue-collar,married,secondary,no,1137,no,no,cellular,6,feb,129,4,211,3,other,no


In [57]:
## metadata of data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4521 entries, 0 to 4520
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        4521 non-null   int64 
 1   job        4521 non-null   object
 2   marital    4521 non-null   object
 3   education  4521 non-null   object
 4   default    4521 non-null   object
 5   balance    4521 non-null   int64 
 6   housing    4521 non-null   object
 7   loan       4521 non-null   object
 8   contact    4521 non-null   object
 9   day        4521 non-null   int64 
 10  month      4521 non-null   object
 11  duration   4521 non-null   int64 
 12  campaign   4521 non-null   int64 
 13  pdays      4521 non-null   int64 
 14  previous   4521 non-null   int64 
 15  poutcome   4521 non-null   object
 16  y          4521 non-null   object
dtypes: int64(7), object(10)
memory usage: 600.6+ KB


Based on the output above, we have -

7 numeric and 10 categorical data
Total of 4521 rows, with no null values. In other words, no missing data.

# *Part One*

*In this part, we will use all the features to train the machine learning models.*

## Data Preparation

In [58]:
## descriptive statistics of the numerical data
data.describe()

,age,balance,day,duration,campaign,pdays,previous
count,4521.000000,4521.000000,4521.000000,4521.000000,4521.000000,4521.000000,4521.000000
mean,41.170095,1422.657819,15.915284,263.961292,2.793630,39.766645,0.542579
std,10.576211,3009.638142,8.247667,259.856633,3.109807,100.121124,1.693562
min,19.000000,-3313.000000,1.000000,4.000000,1.000000,-1.000000,0.000000
25%,33.000000,69.000000,9.000000,104.000000,1.000000,-1.000000,0.000000
50%,39.000000,444.000000,16.000000,185.000000,2.000000,-1.000000,0.000000
75%,49.000000,1480.000000,21.000000,329.000000,3.000000,-1.000000,0.000000
max,87.000000,71188.000000,31.000000,3025.000000,50.000000,871.000000,25.000000


### Data Cleaning

As there are no missing values in the dataset, skipping this step.

### Data Encoding



*   For features that can be ranked, we will use *map()* method to replace feature values.
*   For features that cannot be ranked, we will use *get_dummies()* method to replace feature values.

The purpose of data encoding is to convert categorical data into numerical data so that the machine learning algorithms can process and interpret the data.



In [59]:
"""Ranked Features"""

## Data encoding for education
data['education'] = data['education'].map({'unknown': 0, 'primary': 1, 'secondary': 2, 'tertiary': 3})

## Data encoding for default
data['default'] = data['default'].map({'no': 0, 'yes': 1})

## Data encoding for housing
data['housing'] = data['housing'].map({'no': 0, 'yes': 1})

## Data encoding for loan
data['loan'] = data['loan'].map({'no': 0, 'yes': 1})

## Data encoding for month
data['month'] = data['month'].map({'jan': 1, 'feb': 2, 'mar': 3, 'apr': 4, 'may': 5, 'jun': 6, 'jul': 7, 'aug': 8, 'sep': 9, 'oct': 10, 'nov': 11, 'dec': 12})

## Data encoding for y (target variable)
data['y'] = data['y'].map({'no': 0, 'yes': 1})

In [60]:
"""Un-ranked features"""

from pandas import get_dummies

data = get_dummies(data, columns = ['job', 'marital', 'contact', 'poutcome'], dtype = int)

In [61]:
## check if data has been transformed or not
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4521 entries, 0 to 4520
Data columns (total 35 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   age                4521 non-null   int64
 1   education          4521 non-null   int64
 2   default            4521 non-null   int64
 3   balance            4521 non-null   int64
 4   housing            4521 non-null   int64
 5   loan               4521 non-null   int64
 6   day                4521 non-null   int64
 7   month              4521 non-null   int64
 8   duration           4521 non-null   int64
 9   campaign           4521 non-null   int64
 10  pdays              4521 non-null   int64
 11  previous           4521 non-null   int64
 12  y                  4521 non-null   int64
 13  job_admin.         4521 non-null   int64
 14  job_blue-collar    4521 non-null   int64
 15  job_entrepreneur   4521 non-null   int64
 16  job_housemaid      4521 non-null   int64
 17  job_management

Based on the output above, the data has now a total of 35 columns. This means that the dataset has been encoded properly.

In [62]:
## save the data into a csv file
data.to_csv('bank_encoded.csv')

### Divide the dataset into x (features) and y (target variable)

In [63]:
## divde the data
y = data['y']
x = data.drop('y', axis = 1)

print("Shape of x: ", x.shape)
print("Shape of y: ", y.shape)

Shape of x:  (4521, 34)
Shape of y:  (4521,)


### Data Scaling (only for x)

In this step, we will transform the data into a standard normal distribution using StandardScaler. The purpose of data scaling is to transform the data prepared till now into a format that is suitable for the machine learning model.

In [64]:
## transform the data into standard normal distribution
from sklearn.preprocessing import StandardScaler
from pandas import DataFrame

x_scaled = StandardScaler().fit_transform(x)
DataFrame(x_scaled).to_csv('bank_scaled.csv', index = False)

### Data Splitting

Divide the data into training set, and test set. The purpose of the training set is to make the model learn. Then using the test set, we shall see how well the model learned.

In [65]:
## split the data into training and test set
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x_scaled, y, test_size = 0.25, random_state = 10)

print("x train set: ", x_train.shape)
print("y train set: ", x_test.shape)

print("\nx test set: ", x_test.shape)
print("y test set: ", y_test.shape)

x train set:  (3390, 34)
y train set:  (1131, 34)

x test set:  (1131, 34)
y test set:  (1131,)


### Data Balancing

Using SMOTE, we will create some synthetic data in order to balance the output training set *(y_train)*. The purpose of data balancing step in data preparation is to fix any imbalances in the dataset so as to prevent machine learning models being biased towards classes that have a majority presence.

In [66]:
## Check if y_train has equal number of 0s and 1s
y_train.value_counts()

,count
y,
0,2999
1,391


In [67]:
## values are imbalanced, so balance it using SMOTE

from imblearn.over_sampling import SMOTE

x_train, y_train = SMOTE().fit_resample(x_train, y_train)

y_train.value_counts()

,count
y,
1,2999
0,2999


Now, the dataset is balanced and ready for the next phase, which is Modeling.

## Modeling, Evaluation and Deployment

### Decision Tree Classifier

A Decision Tree Classifier is a supervised ML algorithm that categorizes data by recursively splitting it into smaller sets, till it can surely decide a label.

In [68]:
from sklearn.tree import DecisionTreeClassifier

#### Method 1

##### Modelling

In [69]:
## build the model with max_depth manually entered
dt_classifier = DecisionTreeClassifier(max_depth = 3)

To avoid underfitting or overfitting, train the model using a range of *max_depth* values, later to find a value which makes the model generalised.

In [70]:
## train the model using x_train and y_train dataset
dt_classifier.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=3)

In [71]:
## test the model using x_test
y_pred_dt = dt_classifier.predict(x_test)
y_pred_dt

array([0, 0, 0, ..., 1, 1, 0])

##### Evaluate the model

In [72]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

## Confusion Matrix
conf_matrix_dt = confusion_matrix(y_test, y_pred_dt)
print("Confusion Matrix: \n", conf_matrix_dt)

## Accuracy
accuracy_dt = accuracy_score(y_test, y_pred_dt)
print("\nAccuracy: ", round(accuracy_dt * 100, 2), "%")

## Recall
recall_dt = recall_score(y_test, y_pred_dt)
print("\nRecall: ", round(recall_dt * 100, 2), "%")

## Precision
precision_dt = precision_score(y_test, y_pred_dt)
print("\nPrecision: ", round(precision_dt * 100, 2), "%")

## F1 Score
f1_dt = f1_score(y_test, y_pred_dt)
print("\nF1 Score: ", round(f1_dt * 100, 2), "%")

## deployment decision
if recall_dt > 70 and precision_dt > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")

Confusion Matrix: 
 [[701 300]
 [ 13 117]]

Accuracy:  72.33 %

Recall:  90.0 %

Precision:  28.06 %

F1 Score:  42.78 %

Model not recommended!


#### Method 2

In [73]:
from sklearn.model_selection import GridSearchCV

In [74]:
## define the estimator
dt_classifier_estimator = DecisionTreeClassifier(criterion='entropy')

## define the depth
dt_classifier_depths = {
    'max_depth': [2, 3, 5, 7, 11, 14, 16, 20, 25, 30]
}

## build the model
grid_search_dt_classifier = GridSearchCV(estimator = dt_classifier_estimator, param_grid = dt_classifier_depths, scoring = 'recall', cv = 5)


In [75]:
## train, test, evaluate, and rank
grid_search_dt_classifier.fit(x_scaled, y)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(criterion='entropy'),
             param_grid={'max_depth': [2, 3, 5, 7, 11, 14, 16, 20, 25, 30]},
             scoring='recall')

In [76]:
## best max depth
best_max_depth_dt = grid_search_dt_classifier.best_params_['max_depth']
print("Best Max Depth: ", best_max_depth_dt)

## best recall score
best_recall_dt = grid_search_dt_classifier.best_score_
print("\nBest Recall Score: ", round(best_recall_dt * 100, 2), "%")

## best decision tree
best_dt = grid_search_dt_classifier.best_estimator_
print("\nBest Decision Tree: \n", best_dt)

## deployment decision
if best_recall_dt > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")

Best Max Depth:  25

Best Recall Score:  44.92 %

Best Decision Tree: 
 DecisionTreeClassifier(criterion='entropy', max_depth=25)

Model not recommended!


### Random Forest Classifier

In Random Forest Classifier, instead of building a single decision tree and use it to make predictions, build many slightly different trees and combine their
predictions.

Random Forest Classifier is a type of Ensemble Learning that combines multiple learning algorithms to obtain performance improvements over its components.

In [77]:
from sklearn.ensemble import RandomForestClassifier

#### Method 1

##### Modelling

In [78]:
## build the model
rf_classifier = RandomForestClassifier(criterion = 'entropy', n_estimators = 100)

## train the model using x_train and y_train datasets
rf_classifier.fit(x_train, y_train)

## test the model
y_pred_rf = rf_classifier.predict(x_test)
y_pred_rf

array([0, 0, 0, ..., 0, 0, 0])

To avoid overfitting or underfitting, train the model with a range of *n_estimators* (number of trees), to find the value that generalises the model.

##### Evaluate the model

In [79]:
## Confusion Matrix
conf_matrix_rf = confusion_matrix(y_test, y_pred_rf)
print("Confusion Matrix: \n", conf_matrix_rf)

## Accuracy
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print("\nAccuracy: ", round(accuracy_rf * 100, 2), "%")

## Recall
recall_rf = recall_score(y_test, y_pred_rf)
print("\nRecall: ", round(recall_rf * 100, 2), "%")

## Precision
precision_rf = precision_score(y_test, y_pred_rf)
print("\nPrecision: ", round(precision_rf * 100, 2), "%")

## F1 Score
f1_rf = f1_score(y_test, y_pred_rf)
print("\nF1 Score: ", round(f1_rf * 100, 2), "%")

## decision support
if recall_rf > 70 and precision_rf > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")

Confusion Matrix: 
 [[943  58]
 [ 76  54]]

Accuracy:  88.15 %

Recall:  41.54 %

Precision:  48.21 %

F1 Score:  44.63 %

Model not recommended!


#### Method 2

##### Modelling

In [80]:
## building

rf_classifier_estimator = RandomForestClassifier(criterion = 'entropy')

rf_trees = {
    'n_estimators': [10, 30, 50, 70, 100, 150, 230, 250]
}

grid_search_rf_classifier = GridSearchCV(estimator = rf_classifier_estimator, param_grid = rf_trees, scoring = 'recall', cv = 5)


In [81]:
## train, test, evaluate, and rank

grid_search_rf_classifier.fit(x_scaled, y)

GridSearchCV(cv=5, estimator=RandomForestClassifier(criterion='entropy'),
             param_grid={'n_estimators': [10, 30, 50, 70, 100, 150, 230, 250]},
             scoring='recall')

##### Evaluate the model

In [82]:
## best no of trees
best_rf_trees = grid_search_rf_classifier.best_params_
print("\nBest no. of trees: ", best_rf_trees)

## best recall score
best_recall_rf = grid_search_rf_classifier.best_score_
print("\nBest Recall Score: ", round(best_recall_rf * 100, 2), "%")

best_rf_classifier = grid_search_rf_classifier.best_estimator_
print("\nBest Random Forest Classifier: \n", best_rf_classifier)

if best_recall_rf > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")



Best no. of trees:  {'n_estimators': 50}

Best Recall Score:  22.64 %

Best Random Forest Classifier: 
 RandomForestClassifier(criterion='entropy', n_estimators=50)

Model not recommended!


### Support Vector Machine: Classifier

In [83]:
from sklearn.svm import SVC

#### Method 1

##### Modelling

In [84]:
## build the model
sv_classifier = SVC(kernel = 'linear', C = 1)

In [85]:
## train the model
sv_classifier.fit(x_train, y_train)

SVC(C=1, kernel='linear')

Train the model with a combination of *kernel* and *C* (regularisation paramter) values to get the perfect one that generalises the model.

In [86]:
## test the model
y_pred_svc = sv_classifier.predict(x_test)
y_pred_svc

array([0, 0, 0, ..., 0, 0, 0])

##### Evaluate the model

In [87]:
## Confusion Matrix
conf_matrix_svc = confusion_matrix(y_test, y_pred_svc)
print("Confusion Matrix: \n", conf_matrix_svc)

## Accuracy
accuracy_svc = accuracy_score(y_test, y_pred_svc)
print("\nAccuracy: ", round(accuracy_svc * 100, 2), "%")

## Recall
recall_svc = recall_score(y_test, y_pred_svc)
print("\nRecall: ", round(recall_svc * 100, 2), "%")

## Precision
precision_svc = precision_score(y_test, y_pred_svc)
print("\nPrecision: ", round(precision_svc * 100, 2), "%")

## F1 Score
f1_svc = f1_score(y_test, y_pred_svc)
print("\nF1 Score: ", round(f1_svc * 100, 2), "%")

## deployment decision
if recall_svc > 70 and precision_svc > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")

Confusion Matrix: 
 [[812 189]
 [ 30 100]]

Accuracy:  80.64 %

Recall:  76.92 %

Precision:  34.6 %

F1 Score:  47.73 %

Model not recommended!


#### Method 2

#### Modelling

In [88]:
## build the model

grid_search_sv_classifier = SVC()

kernel_C = {
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
    'C': [0.01, 0.1, 1, 10]
}

grid_search_sv_classifier = GridSearchCV(estimator = grid_search_sv_classifier, param_grid = kernel_C, scoring = 'recall', cv = 5)

In [89]:
## train, test, evaluate, and rank
grid_search_sv_classifier.fit(x_scaled, y)

best_kernel_C = grid_search_sv_classifier.best_params_
print("\nBest Kernel and C: ", best_kernel_C)

recall_svc = grid_search_sv_classifier.best_score_
print("\nBest Recall Score: ", round(recall_svc * 100, 2), "%")

best_svc = grid_search_sv_classifier.best_estimator_
print("\nBest SVC: \n", best_svc)

if recall_svc > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")


Best Kernel and C:  {'C': 10, 'kernel': 'poly'}

Best Recall Score:  33.96 %

Best SVC: 
 SVC(C=10, kernel='poly')

Model not recommended!


# *Part Two*

*In this part, I have used some features (but not all) to train the machine learning models.*

The features that I have used here are ***Age, Job, Education, Balance, Housing, Loan, Poutcome***.

## Data Preparation

In [90]:
## load the data
data2 = read_csv("/content/drive/MyDrive/ML_Dataset/bank.csv")

## remove the columns which is not required
data2 = data2.drop(['marital', 'default', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays', 'previous'], axis=1)
data2

,age,job,education,balance,housing,loan,poutcome,y
0,30,unemployed,primary,1787,no,no,unknown,no
1,33,services,secondary,4789,yes,yes,failure,no
2,35,management,tertiary,1350,yes,no,failure,no
3,30,management,tertiary,1476,yes,yes,unknown,no
4,59,blue-collar,secondary,0,yes,no,unknown,no
...,...,...,...,...,...,...,...,...
4516,33,services,secondary,-333,yes,no,unknown,no
4517,57,self-employed,tertiary,-3313,yes,yes,unknown,no
4518,57,technician,secondary,295,no,no,unknown,no
4519,28,blue-collar,secondary,1137,no,no,other,no


### Data Endcoding

In [91]:
"""Ranked Features"""

## Data encoding for education
data2['education'] = data2['education'].map({'unknown': 0, 'primary': 1, 'secondary': 2, 'tertiary': 3})

## Data encoding for housing
data2['housing'] = data2['housing'].map({'no': 0, 'yes': 1})

## Data encoding for loan
data2['loan'] = data2['loan'].map({'no': 0, 'yes': 1})

## Data encoding for y (target variable)
data2['y'] = data2['y'].map({'no': 0, 'yes': 1})


"""Un-ranked features"""

data2 = get_dummies(data2, columns = ['job', 'poutcome'], dtype = int)



In [92]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4521 entries, 0 to 4520
Data columns (total 22 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   age                4521 non-null   int64
 1   education          4521 non-null   int64
 2   balance            4521 non-null   int64
 3   housing            4521 non-null   int64
 4   loan               4521 non-null   int64
 5   y                  4521 non-null   int64
 6   job_admin.         4521 non-null   int64
 7   job_blue-collar    4521 non-null   int64
 8   job_entrepreneur   4521 non-null   int64
 9   job_housemaid      4521 non-null   int64
 10  job_management     4521 non-null   int64
 11  job_retired        4521 non-null   int64
 12  job_self-employed  4521 non-null   int64
 13  job_services       4521 non-null   int64
 14  job_student        4521 non-null   int64
 15  job_technician     4521 non-null   int64
 16  job_unemployed     4521 non-null   int64
 17  job_unknown   

### Divide the dataset

In [93]:
## divde the data
y2 = data2['y']
x2 = data2.drop('y', axis = 1)

print("Shape of x: ", x2.shape)
print("Shape of y: ", y2.shape)

Shape of x:  (4521, 21)
Shape of y:  (4521,)


### Data Scaling

In [94]:
## transform the data into standard normal distribution
x2_scaled = StandardScaler().fit_transform(x2)
DataFrame(x2_scaled).to_csv('bank_scaled_2.csv', index = False)

### Data Splitting

In [95]:
x2_train, x2_test, y2_train, y2_test = train_test_split(x2_scaled, y2, test_size = 0.25, random_state = 10)

print("x train set: ", x2_train.shape)
print("y train set: ", x2_test.shape)

print("\nx test set: ", x2_test.shape)
print("y test set: ", y2_test.shape)

x train set:  (3390, 21)
y train set:  (1131, 21)

x test set:  (1131, 21)
y test set:  (1131,)


### Data Balancing

In [96]:
## Check if y2_train has equal number of 0s and 1s
y2_train.value_counts()

,count
y,
0,2999
1,391


In [97]:
x2_train, y2_train = SMOTE().fit_resample(x2_train, y2_train)

y2_train.value_counts()

,count
y,
1,2999
0,2999


## Modeling, Evaluation and Deployment

### Decision Tree Classifier

#### Method 1

In [98]:
"""Modeling"""

## build the model with max_depth manually entered
dt_classifier2 = DecisionTreeClassifier(max_depth = 3)

## train the model using x2_train and y2_train dataset
dt_classifier2.fit(x2_train, y2_train)

## test the model using x2_test
y2_pred_dt = dt_classifier2.predict(x2_test)
y2_pred_dt


"""Evaluate and check if model is ready for Deployment"""

## Confusion Matrix
conf_matrix_dt2 = confusion_matrix(y2_test, y2_pred_dt)
print("Confusion Matrix: \n", conf_matrix_dt2)

## Accuracy
accuracy_dt2 = accuracy_score(y2_test, y2_pred_dt)
print("\nAccuracy: ", round(accuracy_dt2 * 100, 2), "%")

## Recall
recall_dt2 = recall_score(y2_test, y2_pred_dt)
print("\nRecall: ", round(recall_dt2 * 100, 2), "%")

## Precision
precision_dt2 = precision_score(y2_test, y2_pred_dt)
print("\nPrecision: ", round(precision_dt2 * 100, 2), "%")

## F1 Score
f1_dt2 = f1_score(y2_test, y2_pred_dt)
print("\nF1 Score: ", round(f1_dt2 * 100, 2), "%")

## deployment decision
if recall_dt2 > 70 and precision_dt2 > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")

Confusion Matrix: 
 [[987  14]
 [110  20]]

Accuracy:  89.04 %

Recall:  15.38 %

Precision:  58.82 %

F1 Score:  24.39 %

Model not recommended!


#### Method 2

In [99]:
"""Modeling"""

## define the estimator
dt_classifier2_estimator = DecisionTreeClassifier(criterion='entropy')

## define the depth
dt_classifier2_depths = {
    'max_depth': [2, 3, 5, 7, 11, 14, 16, 20, 25, 30]
}

## build the model
grid_search_dt_classifier2 = GridSearchCV(estimator = dt_classifier2_estimator, param_grid = dt_classifier2_depths, scoring = 'recall', cv = 5)

## train, test, evaluate, and rank
grid_search_dt_classifier2.fit(x2_scaled, y2)


"""Evaluate and check if model is ready for Deployment"""

## best max depth
best_max_depth_dt2 = grid_search_dt_classifier2.best_params_['max_depth']
print("Best Max Depth: ", best_max_depth_dt2)

## best recall score
best_recall_dt2 = grid_search_dt_classifier2.best_score_
print("\nBest Recall Score: ", round(best_recall_dt2 * 100, 2), "%")

## best decision tree
best_dt2 = grid_search_dt_classifier2.best_estimator_
print("\nBest Decision Tree: \n", best_dt2)

## deployment decision
if best_recall_dt2 > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")

Best Max Depth:  30

Best Recall Score:  24.37 %

Best Decision Tree: 
 DecisionTreeClassifier(criterion='entropy', max_depth=30)

Model not recommended!


### Random Forest Classifier

#### Method 1

In [100]:
"""Modeling"""

## build the model
rf_classifier2 = RandomForestClassifier(criterion = 'entropy', n_estimators = 100)

## train the model using x_train and y_train datasets
rf_classifier2.fit(x2_train, y2_train)

## test the model
y2_pred_rf = rf_classifier2.predict(x2_test)
y2_pred_rf


""""Evaluate and check if model is ready for Deployment"""

## Confusion Matrix
conf_matrix_rf2 = confusion_matrix(y2_test, y2_pred_rf)
print("Confusion Matrix: \n", conf_matrix_rf2)

## Accuracy
accuracy_rf2 = accuracy_score(y2_test, y2_pred_rf)
print("\nAccuracy: ", round(accuracy_rf2 * 100, 2), "%")

## Recall
recall_rf2 = recall_score(y2_test, y2_pred_rf)
print("\nRecall: ", round(recall_rf2 * 100, 2), "%")

## Precision
precision_rf2 = precision_score(y2_test, y2_pred_rf)
print("\nPrecision: ", round(precision_rf2 * 100, 2), "%")

## F1 Score
f1_rf2 = f1_score(y2_test, y2_pred_rf)
print("\nF1 Score: ", round(f1_rf2 * 100, 2), "%")

## decision support
if recall_rf2 > 70 and precision_rf2 > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")

Confusion Matrix: 
 [[884 117]
 [ 90  40]]

Accuracy:  81.7 %

Recall:  30.77 %

Precision:  25.48 %

F1 Score:  27.87 %

Model not recommended!


#### Method 2

In [101]:
"""Modeling"""

## building
rf_classifier2_estimator = RandomForestClassifier(criterion = 'entropy')

rf_trees2 = {
    'n_estimators': [10, 30, 50, 70, 100, 150, 230, 250]
}

grid_search_rf_classifier2 = GridSearchCV(estimator = rf_classifier2_estimator, param_grid = rf_trees2, scoring = 'recall', cv = 5)

## train, test, evaluate, and rank
grid_search_rf_classifier2.fit(x_scaled, y)


"""Evaluate and check if model is ready for Deployment"""

## best no of trees
best_rf_trees2 = grid_search_rf_classifier2.best_params_
print("\nBest no. of trees: ", best_rf_trees2)

## best recall score
best_recall_rf2 = grid_search_rf_classifier2.best_score_
print("\nBest Recall Score: ", round(best_recall_rf2 * 100, 2), "%")

best_rf_classifier2 = grid_search_rf_classifier2.best_estimator_
print("\nBest Random Forest Classifier: \n", best_rf_classifier2)

if best_recall_rf2 > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")


Best no. of trees:  {'n_estimators': 30}

Best Recall Score:  22.65 %

Best Random Forest Classifier: 
 RandomForestClassifier(criterion='entropy', n_estimators=30)

Model not recommended!


### Support Vector Machine: Classifier

#### Method 1

In [102]:
"""Modeling"""

## build the model
sv_classifier2 = SVC(kernel = 'linear', C = 1)

## train the model
sv_classifier2.fit(x2_train, y2_train)

## test the model
y2_pred_svc = sv_classifier2.predict(x2_test)
y2_pred_svc


"""Evaluate and check if model is ready for Deployment"""

## Confusion Matrix
conf_matrix_svc2 = confusion_matrix(y2_test, y2_pred_svc)
print("Confusion Matrix: \n", conf_matrix_svc2)

## Accuracy
accuracy_svc2 = accuracy_score(y2_test, y2_pred_svc)
print("\nAccuracy: ", round(accuracy_svc2 * 100, 2), "%")

## Recall
recall_svc2 = recall_score(y2_test, y2_pred_svc)
print("\nRecall: ", round(recall_svc2 * 100, 2), "%")

## Precision
precision_svc2 = precision_score(y2_test, y2_pred_svc)
print("\nPrecision: ", round(precision_svc2 * 100, 2), "%")

## F1 Score
f1_svc2 = f1_score(y2_test, y2_pred_svc)
print("\nF1 Score: ", round(f1_svc2 * 100, 2), "%")

## deployment decision
if recall_svc2 > 70 and precision_svc2 > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")

Confusion Matrix: 
 [[868 133]
 [ 88  42]]

Accuracy:  80.46 %

Recall:  32.31 %

Precision:  24.0 %

F1 Score:  27.54 %

Model not recommended!


#### Method 2

In [103]:
"""Modeling"""

## build the model
grid_search_sv_classifier2 = SVC()
kernel_C_2 = {
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
    'C': [0.01, 0.1, 1, 10]
}

grid_search_sv_classifier2 = GridSearchCV(estimator = grid_search_sv_classifier2, param_grid = kernel_C_2, scoring = 'recall', cv = 5)


"""Evaluate and check if model is ready for Deployment"""

## train, test, evaluate, and rank
grid_search_sv_classifier2.fit(x2_scaled, y2)

best_kernel_C_2 = grid_search_sv_classifier2.best_params_
print("\nBest Kernel and C: ", best_kernel_C_2)

recall_svc2 = grid_search_sv_classifier2.best_score_
print("\nBest Recall Score: ", round(recall_svc2 * 100, 2), "%")

best_svc2 = grid_search_sv_classifier2.best_estimator_
print("\nBest SVC: \n", best_svc)

if recall_svc2 > 70:
  print("\nModel recommended!")
else:
  print("\nModel not recommended!")


Best Kernel and C:  {'C': 10, 'kernel': 'sigmoid'}

Best Recall Score:  17.47 %

Best SVC: 
 SVC(C=10, kernel='poly')

Model not recommended!
